File: bindings/java/src/main/java/org/indunet/robot/bus/BrokerOptions.java
package org.indunet.robot.bus;

import com.sun.jna.StringArray;
import java.util.Collections;
import java.util.List;

/** Options for starting a {@link Broker} (maps to C {@code RobotBusBrokerOptions}). */
public final class BrokerOptions {
    private final String messageXsubBind;
    private final String messageXpubBind;
    private final String serviceFrontendBind;
    private final String serviceBackendBind;
    private final String actionFrontendBind;
    private final String actionBackendBind;
    private final String grpcListen;
    private final String consoleListen;
    private final boolean tcpOnly;
    private final boolean noConsole;
    private final String brokerId;
    private final List<String> messagePeers;
    private final List<String> servicePeers;
    private final List<String> actionPeers;

    /** Keep-alive for native {@code char**} peer arrays until {@link Broker} start returns. */
    private transient StringArray messagePeersNative;
    private transient StringArray servicePeersNative;
    private transient StringArray actionPeersNative;

    public BrokerOptions() {
        this(null, null, null, null, null, null, null, null, false, false, null, null, null, null);
    }

    public BrokerOptions(
            String messageXsubBind,
            String messageXpubBind,
            String serviceFrontendBind,
            String serviceBackendBind,
            String actionFrontendBind,
            String actionBackendBind,
            String grpcListen,
            String consoleListen,
            boolean tcpOnly,
            boolean noConsole) {
        this(
                messageXsubBind,
                messageXpubBind,
                serviceFrontendBind,
                serviceBackendBind,
                actionFrontendBind,
                actionBackendBind,
                grpcListen,
                consoleListen,
                tcpOnly,
                noConsole,
                null,
                null,
                null,
                null);
    }

    public BrokerOptions(
            String messageXsubBind,
            String messageXpubBind,
            String serviceFrontendBind,
            String serviceBackendBind,
            String actionFrontendBind,
            String actionBackendBind,
            String grpcListen,
            String consoleListen,
            boolean tcpOnly,
            boolean noConsole,
            String brokerId,
            List<String> messagePeers,
            List<String> servicePeers,
            List<String> actionPeers) {
        this.messageXsubBind = messageXsubBind;
        this.messageXpubBind = messageXpubBind;
        this.serviceFrontendBind = serviceFrontendBind;
        this.serviceBackendBind = serviceBackendBind;
        this.actionFrontendBind = actionFrontendBind;
        this.actionBackendBind = actionBackendBind;
        this.grpcListen = grpcListen;
        this.consoleListen = consoleListen;
        this.tcpOnly = tcpOnly;
        this.noConsole = noConsole;
        this.brokerId = brokerId;
        this.messagePeers = messagePeers;
        this.servicePeers = servicePeers;
        this.actionPeers = actionPeers;
    }

    public String getMessageXsubBind() {
        return messageXsubBind;
    }

    public String getMessageXpubBind() {
        return messageXpubBind;
    }

    public String getServiceFrontendBind() {
        return serviceFrontendBind;
    }

    public String getServiceBackendBind() {
        return serviceBackendBind;
    }

    public String getActionFrontendBind() {
        return actionFrontendBind;
    }

    public String getActionBackendBind() {
        return actionBackendBind;
    }

    public String getGrpcListen() {
        return grpcListen;
    }

    public String getConsoleListen() {
        return consoleListen;
    }

    public boolean isTcpOnly() {
        return tcpOnly;
    }

    public boolean isNoConsole() {
        return noConsole;
    }

    public String getBrokerId() {
        return brokerId;
    }

    public List<String> getMessagePeers() {
        return messagePeers == null ? Collections.emptyList() : messagePeers;
    }

    public List<String> getServicePeers() {
        return servicePeers == null ? Collections.emptyList() : servicePeers;
    }

    public List<String> getActionPeers() {
        return actionPeers == null ? Collections.emptyList() : actionPeers;
    }

    RobotBusC.BrokerOptions toNative() {
        RobotBusC.BrokerOptions o = new RobotBusC.BrokerOptions();
        o.messageXsubBind = messageXsubBind;
        o.messageXpubBind = messageXpubBind;
        o.serviceFrontendBind = serviceFrontendBind;
        o.serviceBackendBind = serviceBackendBind;
        o.actionFrontendBind = actionFrontendBind;
        o.actionBackendBind = actionBackendBind;
        o.grpcListen = grpcListen;
        o.consoleListen = consoleListen;
        o.tcpOnly = tcpOnly ? 1 : 0;
        o.noConsole = noConsole ? 1 : 0;
        o.brokerId = brokerId;
        if (messagePeers != null && !messagePeers.isEmpty()) {
            messagePeersNative = new StringArray(messagePeers.toArray(new String[0]));
            o.messagePeers = messagePeersNative;
            o.messagePeerCount = messagePeers.size();
        }
        if (servicePeers != null && !servicePeers.isEmpty()) {
            servicePeersNative = new StringArray(servicePeers.toArray(new String[0]));
            o.servicePeers = servicePeersNative;
            o.servicePeerCount = servicePeers.size();
        }
        if (actionPeers != null && !actionPeers.isEmpty()) {
            actionPeersNative = new StringArray(actionPeers.toArray(new String[0]));
            o.actionPeers = actionPeersNative;
            o.actionPeerCount = actionPeers.size();
        }
        o.write();
        return o;
    }
}

File: bindings/android/src/main/kotlin/org/indunet/robot/bus/BrokerOptions.kt
package org.indunet.robot.bus

import com.sun.jna.StringArray

/** Options for starting a [Broker] (maps to C `RobotBusBrokerOptions`). */
class BrokerOptions
@JvmOverloads
constructor(
    private val messageXsubBind: String? = null,
    private val messageXpubBind: String? = null,
    private val serviceFrontendBind: String? = null,
    private val serviceBackendBind: String? = null,
    private val actionFrontendBind: String? = null,
    private val actionBackendBind: String? = null,
    private val apiListen: String? = null,
    private val consoleListen: String? = null,
    private val tcpOnly: Boolean = false,
    private val noConsole: Boolean = false,
    private val brokerId: String? = null,
    private val messagePeers: List<String>? = null,
    private val servicePeers: List<String>? = null,
    private val actionPeers: List<String>? = null,
    private val noDiscovery: Boolean = false,
    private val domainId: Int = 0,
    private val advertiseHost: String? = null,
    private val peers: List<String>? = null,
    private val noTank: Boolean = false,
) {
    /** Keep-alive for native `char**` peer arrays until [Broker] start returns. */
    private var messagePeersNative: StringArray? = null
    private var servicePeersNative: StringArray? = null
    private var actionPeersNative: StringArray? = null
    private var peersNative: StringArray? = null

    fun getMessageXsubBind(): String? = messageXsubBind

    fun getMessageXpubBind(): String? = messageXpubBind

    fun getServiceFrontendBind(): String? = serviceFrontendBind

    fun getServiceBackendBind(): String? = serviceBackendBind

    fun getActionFrontendBind(): String? = actionFrontendBind

    fun getActionBackendBind(): String? = actionBackendBind

    fun getApiListen(): String? = apiListen

    fun getConsoleListen(): String? = consoleListen

    fun isTcpOnly(): Boolean = tcpOnly

    fun isNoConsole(): Boolean = noConsole

    fun isNoTank(): Boolean = noTank

    fun getBrokerId(): String? = brokerId

    fun getMessagePeers(): List<String> = messagePeers.orEmpty()

    fun getServicePeers(): List<String> = servicePeers.orEmpty()

    fun getActionPeers(): List<String> = actionPeers.orEmpty()

    fun getPeers(): List<String> = peers.orEmpty()

    fun isNoDiscovery(): Boolean = noDiscovery

    fun getDomainId(): Int = domainId

    fun getAdvertiseHost(): String? = advertiseHost

    internal fun toNative(): RobotBusC.BrokerOptions {
        val o = RobotBusC.BrokerOptions()
        o.messageXsubBind = messageXsubBind
        o.messageXpubBind = messageXpubBind
        o.serviceFrontendBind = serviceFrontendBind
        o.serviceBackendBind = serviceBackendBind
        o.actionFrontendBind = actionFrontendBind
        o.actionBackendBind = actionBackendBind
        o.apiListen = apiListen
        o.consoleListen = consoleListen
        o.tcpOnly = if (tcpOnly) 1 else 0
        o.noConsole = if (noConsole) 1 else 0
        o.brokerId = brokerId
        if (!messagePeers.isNullOrEmpty()) {
            messagePeersNative = StringArray(messagePeers.toTypedArray())
            o.messagePeers = messagePeersNative
            o.messagePeerCount = messagePeers.size.toLong()
        }
        if (!servicePeers.isNullOrEmpty()) {
            servicePeersNative = StringArray(servicePeers.toTypedArray())
            o.servicePeers = servicePeersNative
            o.servicePeerCount = servicePeers.size.toLong()
        }
        if (!actionPeers.isNullOrEmpty()) {
            actionPeersNative = StringArray(actionPeers.toTypedArray())
            o.actionPeers = actionPeersNative
            o.actionPeerCount = actionPeers.size.toLong()
        }
        o.noDiscovery = if (noDiscovery) 1 else 0
        o.domainId = domainId
        o.advertiseHost = advertiseHost
        if (!peers.isNullOrEmpty()) {
            peersNative = StringArray(peers.toTypedArray())
            o.peers = peersNative
            o.peerCount = peers.size.toLong()
        }
        o.noTank = if (noTank) 1 else 0
        o.write()
        return o
    }
}
